### ISSUE TITLE
Compiler bug: Could not find type of...

### ISSUE THREAD

**Issue by jnfoster** [p4lang/p4c#817]
Compiling this program,
```
#include <core.p4>

// Architecture    
control C();
package S(C c);
extern BoolRegister {
  BoolRegister();
  bool get();
  void flip();
}

// User Program
BoolRegister() r;
    
action a(in C c1, in C c2) {
  r.flip();
  if(r.get()) {
    c1.apply();
  } else {
    c2.apply();
  }
}

control MyC1() {
  apply {
  }
}
control MyC2() {
  apply {    
  }
}
control MyC3() {
  MyC1() c1;
  MyC2() c2;    
  apply {
    a(c1, c2);
  }
}

S(MyC3()) main;
```
crashes with a compiler bug:
```
In file: p4c/frontends/p4/typeChecking/typeChecker.h:101
Compiler Bug: cyclic.p4(33): Could not find type of @name("c1") MyC1 c1()
  MyC1() c1;
         ^^
```

**Comment by jafingerhut** [comment]
I'm not saying that the compiler should crash, but I have a question raised by your example program.

Do you think it should be legal to call apply() on a control from inside of an action?  I would have guessed that this isn't legal P4_16 at all, at least if that action is invoked as the action of a table, as in my example program below.

If it is legal P4_16, should this be allowed in PSA portable programs?

The example program below gives no errors from the latest version of p4test and p4c-bm2-ss as of 2017-Jul-19:
```
#include <core.p4>

// Architecture    
control C();
package S(C c);

// User Program
control MyC1() {
    action b() { }
    table t1 {
        key = { }
        actions = { b; }
    }
    apply {
        t1.apply();
    }
}
control MyC3() {
    MyC1() c1;
    action a() {
        c1.apply();
    }
    table t2 {
        key = { }
        actions = { a; }
    }
    apply {
        t2.apply();
    }
}
S(MyC3()) main;
```

**Comment by mbudiu-vmw** [comment]
This issue is fixed by PR #820.
@jafingerhut has a very good question here, but that's a separate issue.
I think that actions should not be allowed to call controls.

**Comment by jafingerhut** [comment]
I have opened an issue on the p4lang/p4-spec repository, as a more appropriate place for my question about what things actions are allowed to call: https://github.com/p4lang/p4-spec/issues/368

Feel free to disregard that question of mine on this issue, as far as determining when this one ought to be closed.
